### DISCUSSION TITLE
Duplicate pipelines for GitLab merged results pipelines with minimumReleaseAge

### DISCUSSION THREAD

**Discussion by mikkopiu** [renovatebot/renovate#23278] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.4.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

When using the GitLab Premium feature [Merged results pipeline](https://docs.gitlab.com/16.1/ee/ci/pipelines/merged_results_pipelines.html), GitLab will create a new `HEAD` commit on a Merge Request's branch and therefore Renovate will not be adding its statuses, like `renovate/stability-days` (`minimumReleaseAge`/`stabilityDays`), to the correct commit from GitLab's perspective.

This causes two main problems:

1. There will be duplicate pipelines: one for the actual commit with Renovate's status(es) and one _merged results_ pipeline by GitLab for the internal merge commit (example in related discussion: https://github.com/renovatebot/renovate/discussions/23132)
2. With GitLab currently not handling multiple _different_ pipeline type statuses for a single Merge Request, only one of the pipelines will be considered for the MR's own status (and subsequently, the "pipelines must succeed" rule) -> one pipeline status will inevitably be ignored

This issue is made worse by the experimental variable, `RENOVATE_X_GITLAB_BRANCH_STATUS_DELAY`, as it might cause the Renovate single-job pipeline to be the latest one for an MR -> people can accidentally merge code that would have broken in the "real" pipeline (the merged results pipeline).

The solution would probably have to be similarly to the automerge solution where Renovate would:

1. Get the Merge Request's `.head_pipeline.sha`
2. If exists -> use that commit instead of the one just created
3. If doesn't exist -> use branch ref, like currently

I'm not familiar with GitHub's merge queues and whether they work the same way and whether Renovate already handles them (couldn't find anything in sources, at least).

Related issues/PRs:

- https://github.com/renovatebot/renovate/issues/17440 - AFAICT this only added support for merged results pipelines in relation to Renovate's automerging, and doesn't help here
- https://github.com/renovatebot/renovate/issues/5573 - Possibly related on a conceptual level
- https://github.com/renovatebot/renovate/discussions/23132

As this is a GitLab Premium feature and I don't have access to GitLab.com group with Premium (only self-hosted), I can't create a minimal repro.

### Relevant debug logs

<details><summary>Logs</summary>

Commit push:

```json
{"name":"renovate","hostname":"runnername","pid":34,"level":20,"logContext":"0AF5wn5uCFNbP3_KzFM0E","repository":"reponame","branch":"renovate/apollo-graphql-packages","msg":"Pushing refSpec renovate/apollo-graphql-packages:renovate/apollo-graphql-packages","time":"2023-07-10T07:48:37.153Z","v":0}
```

Status check starts (**NOTE:** This is _before_ the configured `RENOVATE_X_GITLAB_BRANCH_STATUS_DELAY=5000` delay)

```json
{"name":"renovate","hostname":"runnername","pid":34,"level":20,"logContext":"0AF5wn5uCFNbP3_KzFM0E","repository":"reponame","branch":"renovate/apollo-graphql-packages","msg":"Updating renovate/stability-days status check state to green","time":"2023-07-10T07:48:38.242Z","v":0}
```

`2023-07-10T07:48:40.212Z` - Merged results pipeline created, this is now `head_pipeline` for MR

`2023-07-10T07:48:43.337Z` - Renovate's branch pipeline created, changes to `head_pipeline` for MR and causes Merged results pipeline to be effectively ignored

Next log event (shows 5s delay)

```json
{"name":"renovate","hostname":"runnername","pid":34,"level":20,"logContext":"0AF5wn5uCFNbP3_KzFM0E","repository":"reponame","branch":"renovate/apollo-graphql-packages","msg":"Ensuring PR","time":"2023-07-10T07:48:43.579Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by mikkopiu** [comment]
Tested GitLab's side of things:

1. Push commit & let GitLab start a Merged Results pipeline
2. Get `.head_pipeline.sha` from `/api/v4/projects/:id/merge_requests/:iid`
3. POST new status to `/api/v4/projects/:id/statuses/:sha?state=success&name=sometestname&description=from-curl`

-> shows up in the correct pipeline, compared to setting the status on the actual commit, before or after the merged results pipeline is started which will show up as its own pipeline.

**Comment by rarkins** [comment]
We also need to think about the implications of if the user has Renovate option `prCreation` set to anything except "immediate", because then it _must_ set statuses on the "real" commit and not the merge result commit, prior to creating the MR.

**Reply by mikkopiu** [reply]
That's true 🤔 And there might be additional complications from users changing `prCreation` settings during an MR's lifetime (though, I'm not sure how well-covered those use cases are currently).

**Reply by rarkins** [reply]
If Renovate makes a commit, adds some commit status, then creates the MR, then adds same status to the merge result commit, then what is the impact? Right now Renovate sets status always after commit creation.

**Reply by mikkopiu** [reply]
I believe it would then show up as a branch pipeline (and assuming any possible slowness of GitLab is handled with `RENOVATE_X_GITLAB_BRANCH_STATUS_DELAY` or the default 1s delay, it would be the _same_ pipeline as any _possibly_ configured branch pipelines for a project) **but** would mean that the status wouldn't show up in the merged results pipeline (and therefore the MR's status overview thingy) -- which can be fine if Renovate's status jobs aren't necessary to be visible to users (users would have to know which pipeline to check).

That could be one option, with the visibility caveat.
